Backports release 1.6.1

KristofferC opened the pull request:
Backported PRs:
- [x] JuliaLang/julia#39372 - Fixed a spelling error.
- [x] JuliaLang/julia#39891 - Add new LLVM patches for Aarch64 issues
- [x] JuliaLang/julia#40095 - fix JuliaLang/julia#40050, handling fields that are pointers due to subtype circularity
- [x] JuliaLang/julia#38981 - Fix JuliaLang/julia#38491: fix an abspath() edge case on Windows
- [x] JuliaLang/julia#40163 - Have different Windows uninstall keys for different Julia versions
- [x] JuliaLang/julia#40181 - README: git checkout 1.6.0 in the instructions.
- [x] JuliaLang/julia#40176 - Fix armv7l trampoline
- [x] JuliaLang/julia#40174 - Remove eltype assumption from ldiv!(LowerTriangular(SparseMatrixCSC{Tv,Ti), ...)
- [x] JuliaLang/julia#40208 - Fix a small bug in the TOML parser and add an example of printing to a file
- [x] JuliaLang/julia#40161 - Fix issue where approved credentials were not sent to git credential helpers
- [x] JuliaLang/julia#40203 - fix markdown list rendering
- [x] JuliaLang/julia#40264 - Make some constructors bold in stacktraces
- [x] JuliaLang/julia#40275 - Add aliasscope expr to interpreter
- [x] JuliaLang/julia#40281 - add a missing propagate_inbounds to a getindex method
- [x] JuliaLang/julia#40274 - replace broken replaceuserpath function with existing contractuser one
- [x] JuliaLang/julia#40321 - Fix rounding cases for fixed-width printing in ryu routines
- [x] JuliaLang/julia#40323 - Fix relpath when path and startpath are in the same drive
- [x] JuliaLang/julia#40280 - exempt begin and underscore from macro hygiene
- [x] JuliaLang/julia#40293 - fix macro hygiene for `@nospecialize(::T)`
- [x] JuliaLang/julia#39310 - Update PCRE2 to 10.36
- [x] JuliaLang/julia#40261 - fix JuliaLang/julia#40258: nested string interpolation
- [x] JuliaLang/julia#40319 - Fix JuliaLang/julia#40318 by using UInt to convert instead of Int on Ptr
- [x] JuliaLang/julia#40294 - Fix reduce(vcat,...) type inference (#40277)
- [x] JuliaLang/julia#40286 - fix JuliaLang/julia#40249, reshaping SharedArray on another process
- [x] JuliaLang/julia#40217 - [build] Add `symlink_system_library` calls for CSL
- [x] JuliaLang/julia#40395 - fix JuliaLang/julia#40337, no error info from serialized TaskFailedException
- [x] JuliaLang/julia#40403 - Fix temp install of BugReporting.jl, fixes JuliaLang/julia#40398.
- [x] JuliaLang/julia#40428 - fix a typo in TOML error message
- [x] JuliaLang/julia#40364 - searchsorted ranges: simplify code, fix bug with Unsigned needle
- [x] JuliaLang/julia#40365 - fix JuliaLang/julia#32843, codegen attempting `typeof(NULL)` for some union tags
- [x] JuliaLang/julia#40421 -  Bump the Tar stdlib from ac4d442 to 86825ec 


Non-merged PRs with backport label:
- [ ] JuliaLang/julia#40038 - Allow the conversion of AbstractUnitRanges to OrdinalRanges
- [ ] JuliaLang/julia#39544 - fix definition of open I/O to Process

omus committed:
Update PCRE2 to 10.36 (#39310)

(cherry picked from commit c889de7eb80d42f745652da84b99c127a581d0ff)

motchy869 committed:
Fixed a spelling error. (#39372)

(cherry picked from commit cc4690a6a746cac3961a59ac8026f57d6d52351c)

Keno committed:
Add new LLVM patches for Aarch64 issues (#39891)

Fixes #39818
Fixes #39820
Fixes #39823

(cherry picked from commit 0988fcfb2b79416139343b3ca48e01b5a7be8120)

JeffBezanson committed:
fix #40050, handling fields that are pointers due to subtype circularity (#40095)

(cherry picked from commit e3dffb9da5e2af961ccebf3d82de138bbda46d4d)

belamenso committed:
Fix #38491: fix an abspath() edge case on Windows (#38981)

* Fix #38491: fix an abspath() edge case on Windows

* Update base/path.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

* Update base/path.jl

* Update test/path.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Co-authored-by: Mustafa M <mus-m@outlook.com>
Co-authored-by: Jameson Nash <vtjnash@gmail.com>
(cherry picked from commit f023677ce8b22a266b520b5b660114ccfce655e3)

jaakkor2 committed:
Have different Windows uninstall keys for different Julia versions (#40163)

* Have different Windows uninstall keys for different Julia versions

Should fix #40004.  Have the uninstall key to be "{#AppNameLong}_is1" where AppNameLong contains the Julia version to prevent Windows from overwriting the uninstall information for different Julia versions.

Inno Setup documentation about the AppId https://jrsoftware.org/ishelp/topic_setup_appid.htm copied here for convenience
```
[Setup]: AppId

Default value:

AppName
Description:

The value of AppId is stored inside uninstall log files (unins???.dat), and is checked by subsequent installations to determine whether it may append to a particular existing uninstall log. Setup will only append to an uninstall log if the AppId of the existing uninstall log is the same as the current installation's AppId. For a practical example, say you have two installations -- one entitled My Program and the other entitled My Program 1.1 Update. To get My Program 1.1 Update to append to My Program's uninstall log, you would have to set AppId to the same value in both installations.

AppId also determines the actual name of the Uninstall registry key, to which Inno Setup tacks on "_is1" at the end. (Therefore, if AppId is "MyProgram", the key will be named "MyProgram_is1".) Pre-1.3 versions of Inno Setup based the key name on the value of AppVerName.

AppId is a not used for display anywhere, so feel free to make it as cryptic as you desire. The value may include constants.

If you use a {code:..} constant to allow your user to customize AppId, you do not need to return the real value until just before the installation starts: if necessary you may return an empty or generic value at earlier times. If not empty, this value will only be used to attempt to restore previous install settings (like the settings stored by [Setup] section directive UsePreviousAppDir). If empty, it isn't used for anything.

The length of AppId with all constants evaluated should never exceed 127 characters.
Example:

AppId=MyProgram
```

* Use DirName as AppId

Co-authored-by: Mustafa M <mus-m@outlook.com>

* remove unneeded code

Co-authored-by: Mustafa M <mus-m@outlook.com>
(cherry picked from commit e2f5f26aec6e0a3e239d3382aa7007c3caf85f5e)

fredrikekre committed:
README: git checkout 1.6.0 in the instructions.

(cherry picked from commit c6d5c406974268c6f402ce036da4c0d70b849fd0)

staticfloat committed:
Fix armv7l trampoline (#40176)

We copy-pasted the wrong macro here, should have used `CNAMEADDR()` not `CNAME()`.  This causes segfaults on armv7l processors by trying to read from an incorrect address when loading the trampoline target address value.

(cherry picked from commit 1897e08ff54eb8c91b806e20bbfecbe7b2d4f9f7)

palday committed:
Remove eltype assumption from ldiv!(LowerTriangular(SparseMatrixCSC{Tv,Ti), ...) (#40174)

(cherry picked from commit beeb6d776678000c32cd3992d0bc4a65fe1aa5e3)

KristofferC committed:
Fix a small bug in the TOML parser and add an example of printing to a file (#40208)

fix the TOML parser to parse 0e-3 and add an example of writing to a file

(cherry picked from commit 2ec75d65ce13651f42f59456568aefe30901b9c6)

omus committed:
Fix issue where approved credentials were not sent to git credential helpers (#40161)

Due to the bug in the `GitCredential` constructor we were never writing
credential data to the credential helper effectively making our use of
git credential helpers read-only.

(cherry picked from commit ac50ac6058caa97c047c1751ca9cbbe0c1ceea2a)

KristofferC committed:
fix markdown list rendering (#40203)

(cherry picked from commit 0f828a2acc497f0c26965956a57cfa9ea69cdd09)

mcabbott committed:
make some constructors bold in stacktraces (#40264)

(cherry picked from commit 42766ace9af350fd51cd33d0890fe6bb66a1638b)

mkitti commented:
I have a bugfix for https://github.com/JuliaLang/julia/issues/39798 in https://github.com/JuliaLang/julia/pull/39808. Minor change.

simeonschaub committed:
exempt begin and underscore from macro hygiene (#40280)

(cherry picked from commit a52b43d04e7e5044d78858d7e8a374b1f181f138)

vchuravy committed:
Add aliasscope expr to interpreter (#40275)

(cherry picked from commit 8868d3fb15eed14de477c9d77b24e40e05219c07)

simeonschaub committed:
fix #40258: nested string interpolation (#40261)

* fix #40258: nested string interpolation

fixes #40258

* fix interpolation with varargs

* add additional test case

* add additional test cases

Co-authored-by: Nathan Daly <NHDaly@gmail.com>

* keep previous behavior, don't try to be too smart

Co-authored-by: Nathan Daly <NHDaly@gmail.com>
Co-authored-by: Jeff Bezanson <jeff.bezanson@gmail.com>
(cherry picked from commit 637f52b8eb836bda45bdb282ff3dd3d347bef439)

KristofferC committed:
add a missing propagate_inbounds to a getindex method (#40281)

* add a missing propagate_inbounds

(cherry picked from commit 80002db74c88fae5803688e930cd77ce0567c355)

vtjnash committed:
replace broken replaceuserpath function with existing contractuser one (#40274)

Fixes #39674

(cherry picked from commit d07b5a134632d26bdbadde1b0352d2de8aae2906)

quinnj committed:
Fix #40318 by using UInt to convert instead of Int on Ptr

(cherry picked from commit 0d9163b561a5f809b895b9cd8893d1a010197284)

quinnj committed:
Fix rounding cases for fixed-width printing in ryu routines

Fixes #40303. When printing values to fixed widths through the
`Ryu.writefixed` or `Ryu.writeexp` routines, we have a "cleanup" section
after a value has been printed to see if it needs to be rounded given
the input precision and width. The core issue was the terminating
condition: it previously only checked if we were at the start of a
buffer or had encountered the `'-'` character. Via Printf formatting,
however, it also allows specifying the `'+'` and `' '` characters to
preceed a formatted number. Hence, in the OP, the `'1'` character was
getting "rounded" up to the `','` character. The fix here is correctly
checking if the `plus` or `space` options were passed to the routine and
if so, include those in our rounding termination check. The original
issue only reported the "plus" case for the `f` format specifier, but
the same bug affects the `e` format specifier and the "space" option.

(cherry picked from commit fcf6779b1ec119b7a3686d08dfba7ba0acfd3f93)

quinnj committed:
Fix 32-bit

(cherry picked from commit 5332d7d4e40ca2189e4ab597bf937f272f471bda)

simeonschaub committed:
fix macro hygiene for `@nospecialize(::T)` (#40293)

(cherry picked from commit 73aa7e4df5775dce521615a0da5c09e1735a3cd2)

dlfivefifty committed:
Fix reduce(vcat,...) type inference (#40277) (#40294)

(cherry picked from commit 03abc458ea73bfa14713a6362fc60b097e7be337)

musm committed:
Fix relpath when path and startpath are in the same drive (#40323)

* Fix relpath when path and startpath are in the same drive

When startpath == ".", assume the startpath is in the same drive.
This subsequently required tweaking the existing logic to then
canonicalize the drive casing in instances where the drive casing
differs.

(cherry picked from commit 89fff18d590db9be84598e0045003b903d48d2df)

vtjnash committed:
searchsorted ranges: simplify code, fix bug with Unsigned needle

This avoids computing `length` unnecessarily, which often requires a
division, and is therefore often expensive.

(cherry picked from commit 40d4f4a0d71b3129282e58b9a2f9b7f930ad13d3)

vtjnash committed:
argmin ranges bug: you cannot pass an unsorted array to searchsorted

(cherry picked from commit 454fc51714c9ba2a5122268fb9405e6f6f4bfc11)

JeffBezanson committed:
fix #32843, codegen attempting `typeof(NULL)` for some union tags

also fixes part of #40065

(cherry picked from commit 00cf7b0542c4b6f6e5e4d5b6d8ce81a75693fedd)

martinholters committed:
Normalize (simplify) `UnionAll`s when used as type parameter (#36211)

(cherry picked from commit fedefe913a265581d6ba4444b8bcd6dbf77e333b)

JeffBezanson committed:
fix #40249, reshaping SharedArray on another process (#40286)

(cherry picked from commit 79e198b29f620114e86950cf9f24c48b049d9c67)

staticfloat committed:
[build] Add `symlink_system_library` calls for CSL (#40217)

* [build] Add `symlink_system_library` calls for CSL

We differentiate between a `USE_BINARYBUILDER_CSL=0` configuration
(where we copy the CSL libraries into the Julia build tree) versus a
`USE_SYSTEM_CSL=1` configuration (where we symlink them in a manner
similar to other system libararies)

* Expand `JL_PRIVATE_LIBS-1` for versioned library names as well

Without this, we fail to install things like `libgcc_s.so.1` when
installed into the private libdir by `symlink_system_libraries`.

(cherry picked from commit 336cdf80f38b89b42e2fb72e1057f6fced1493d0)

BenjaminGalliot committed:
Backport 1.6 of PR #40328. (#40377)

JeffBezanson committed:
fix #40337, no error info from serialized TaskFailedException (#40395)

(cherry picked from commit cc1e70a6e7d53e64c3769b60de196dd332a06584)

fredrikekre committed:
Fix temp install of BugReporting.jl (#40403)

Fixes #40398

(cherry picked from commit f021c67fe9aba889f015b66c6d5bf6394762c60d)

DilumAluthgeBot committed:
[automated] Bump the Tar stdlib from ac4d442 to 86825ec (#40421)

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>
(cherry picked from commit 2f0156a6c79566bf265397641effa17a202f7ab4)

KristofferC committed:
fix a typo in TOML error message (#40428)

* fix a typo in TOML error message

(cherry picked from commit e260f2bccada3ef162c1f322bd80581a3d95fb64)

KristofferC commented:
@nanosoldier `runtests(ALL, vs = ":release-1.6")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/40209#issuecomment-817364568):
```
NanosoldierError: failed to run tests: TaskFailedException:
failed process: Process(`docker stop ONNX-5R9QKjsa`, ProcessExited(1)) [1]

pipeline_error at ./process.jl:525 [inlined]
#run#596 at ./process.jl:440
run at ./process.jl:438 [inlined]
kill_container at /storage/pkgeval/dev/PkgEval/src/run.jl:364 [inlined]
stop at /storage/pkgeval/dev/PkgEval/src/run.jl:167 [inlined]
#36 at /storage/pkgeval/dev/PkgEval/src/run.jl:234 [inlined]
lock at ./lock.jl:161
macro expansion at /storage/pkgeval/dev/PkgEval/src/run.jl:231 [inlined]
#35 at ./task.jl:356
wait at ./task.jl:267 [inlined]
fetch at ./task.jl:282 [inlined]
#run_sandboxed_test#30 at /storage/pkgeval/dev/PkgEval/src/run.jl:258
macro expansion at /storage/pkgeval/dev/PkgEval/src/run.jl:584 [inlined]
#50 at ./task.jl:356
```
Logs and partial data can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/f2bc4d3_vs_9283b6c/report.md)
cc @maleadt

KristofferC commented:
@nanosoldier `runtests(ALL, vs = ":release-1.6")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/40209#issuecomment-817538111) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/f2bc4d3_vs_9283b6c/report.md). cc @maleadt

KristofferC commented:
@nanosoldier `runtests(["ADNLPModels", "AnimalBreedingTools", "AxisKeys", "BifurcationKit", "DistributedOperations", "FameSVD", "Hecke", "ImageProjectiveGeometry", "IntervalTrees", "JuliaInterpreter", "Lale", "MINLPTests", "MPIReco", "MadNLP", "MusicManipulations", "NODAL", "NiceNumbers"], vs = ":release-1.6")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/40209#issuecomment-817764958) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/f2bc4d3_vs_9283b6c/report.md). cc @maleadt

KristofferC committed:
bump Pkg to 1.6.1

KristofferC commented:
AxisKeys is https://github.com/mcabbott/AxisKeys.jl/issues/57.

KristofferC committed:
prevent conj! into uninitialized memory (#40481)

(cherry picked from commit aa7d8797223422a31573b0593ce7267b6c0ea63a)

KristofferC commented:
@nanosoldier `runtests(ALL, vs = ":release-1.6")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/40209#issuecomment-820266565) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/23e5f6f_vs_9283b6c/report.md). cc @maleadt

KristofferC commented:
@nanosoldier `runtests(["AxisKeys", "BioMASS", "BlobTracking", "CancerImagingArchive", "Circuitscape", "DeepForest", "EliminateGraphs", "FastFloat16s", "FunctionOperators", "Graph500", "GtkObservables", "JSOSolvers", "SignalingDimension", "SimpleTypePrint"], vs = ":release-1.6")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/40209#issuecomment-821310377) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/23e5f6f_vs_9283b6c/report.md). cc @maleadt

KristofferC commented:
@nanosoldier `runbenchmarks(ALL, vs = ":release-1.6")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/40209#issuecomment-821524456) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/58ac831_vs_9283b6c/report.md). cc @christopher-dG
